#### Importing necessary libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
from sklearn.model_selection import train_test_split

C:\Users\Vipra\anaconda3\envs\DS-Inno\Lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\Vipra\anaconda3\envs\DS-Inno\Lib\site-packages\numpy\.libs\libopenblas.FB5AE2TYXYH2IJRDKGDGQ3XBKLKTF43H.gfortran-win_amd64.dll
C:\Users\Vipra\anaconda3\envs\DS-Inno\Lib\site-packages\numpy\.libs\libopenblas64__v0.3.21-gcc_10_3_0.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


#### Importing dataset

In [2]:
df = pd.read_csv(r"uber_rides_data.xlsx - sample_train.csv")

In [3]:
### Finding out the shape of the dataset
df.shape

(200000, 8)

In [5]:
### Getting the info about the dataframe and its features
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 8 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   ride_id            200000 non-null  int64  
 1   fare_amount        200000 non-null  float64
 2   pickup_datetime    200000 non-null  object 
 3   pickup_longitude   200000 non-null  float64
 4   pickup_latitude    200000 non-null  float64
 5   dropoff_longitude  199999 non-null  float64
 6   dropoff_latitude   199999 non-null  float64
 7   passenger_count    200000 non-null  int64  
dtypes: float64(5), int64(2), object(1)
memory usage: 12.2+ MB


#### Number of integer columns in dataset are 2

In [10]:
### Statistical description of numerical columns
df.describe()

ride_id    fare_amount  pickup_longitude  pickup_latitude  \
count  1.999990e+05  199999.000000     199999.000000    199999.000000   
mean   2.771248e+07      11.359892        -72.527631        39.935881   
std    1.601386e+07       9.901760         11.437815         7.720558   
min    1.000000e+00     -52.000000      -1340.648410       -74.015515   
25%    1.382534e+07       6.000000        -73.992065        40.734796   
50%    2.774524e+07       8.500000        -73.981823        40.752592   
75%    4.155535e+07      12.500000        -73.967154        40.767158   
max    5.542357e+07     499.000000         57.418457      1644.421482   

       dropoff_longitude  dropoff_latitude  passenger_count  
count      199999.000000     199999.000000    199999.000000  
mean          -72.525292         39.923890         1.684543  
std            13.117408          6.794829         1.385995  
min         -3356.666300       -881.985513         0.000000  
25%           -73.991407         40.733823         1.000000  
50%           -73.980093         40.753042         1.000000  
75%           -73.963659         40.768001         2.000000  
max          1153.572603        872.697628       208.000000

In [7]:
### Finding null values int he dataframe
df.isna().sum()

ride_id              0
fare_amount          0
pickup_datetime      0
pickup_longitude     0
pickup_latitude      0
dropoff_longitude    1
dropoff_latitude     1
passenger_count      0
dtype: int64

#### number of missing values in dropoff_longitude is 1

In [8]:
### dropping the null values
df.dropna(inplace = True)

In [56]:
### Finding statistical description after removing the null values
df.describe()

ride_id    fare_amount  pickup_longitude  pickup_latitude  \
count  1.999990e+05  199999.000000     199999.000000    199999.000000   
mean   2.771248e+07      11.359892        -72.527631        39.935881   
std    1.601386e+07       9.901760         11.437815         7.720558   
min    1.000000e+00     -52.000000      -1340.648410       -74.015515   
25%    1.382534e+07       6.000000        -73.992065        40.734796   
50%    2.774524e+07       8.500000        -73.981823        40.752592   
75%    4.155535e+07      12.500000        -73.967154        40.767158   
max    5.542357e+07     499.000000         57.418457      1644.421482   

       dropoff_longitude  dropoff_latitude  passenger_count  \
count      199999.000000     199999.000000    199999.000000   
mean          -72.525292         39.923890         1.684543   
std            13.117408          6.794829         1.385995   
min         -3356.666300       -881.985513         0.000000   
25%           -73.991407         40.733823         1.000000   
50%           -73.980093         40.753042         1.000000   
75%           -73.963659         40.768001         2.000000   
max          1153.572603        872.697628       208.000000   

       haversine_distance  ride_week_day  
count       199999.000000  199999.000000  
mean            20.855350       3.048435  
std            382.964642       1.946946  
min              0.000000       0.000000  
25%              1.215222       1.000000  
50%              2.120992       3.000000  
75%              3.875169       5.000000  
max          16409.239135       6.000000

#### Average fare value  is 11.359892

#### Converting the pickup_datatime column to 'datatime' type

In [14]:
df['pickup_datetime'] = df['pickup_datetime'].str.replace("UTC","")

In [16]:
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])

### Haversine distance formula

In [17]:
import math
 
def haversine(lat1, lon1, lat2, lon2):
 
    dLat = (lat2 - lat1) * math.pi / 180.0
    dLon = (lon2 - lon1) * math.pi / 180.0
 
    lat1 = (lat1) * math.pi / 180.0
    lat2 = (lat2) * math.pi / 180.0
 
    # apply formulae
    a = (pow(math.sin(dLat / 2), 2) +
         pow(math.sin(dLon / 2), 2) *
             math.cos(lat1) * math.cos(lat2));
    rad_earth = 6371
    c = 2 * math.asin(math.sqrt(a))
    return rad_earth * c


In [21]:
df_temp = df[['pickup_latitude','pickup_longitude','dropoff_latitude','dropoff_longitude']]

In [22]:
df_temp

pickup_latitude  pickup_longitude  dropoff_latitude  dropoff_longitude
0             40.738354        -73.999817         40.723217         -73.999512
1             40.728225        -73.994355         40.750325         -73.994710
2             40.740770        -74.005043         40.772647         -73.962565
3             40.790844        -73.976124         40.803349         -73.965316
4             40.744085        -73.925023         40.761247         -73.973082
...                 ...               ...               ...                ...
199995        40.739367        -73.987042         40.740297         -73.986525
199996        40.736837        -73.984722         40.739620         -74.006672
199997        40.756487        -73.986017         40.692588         -73.858957
199998        40.725452        -73.997124         40.695416         -73.983215
199999        40.720077        -73.984395         40.768793         -73.985508

[199999 rows x 4 columns]

In [24]:
df['haversine_distance'] = df_temp.apply(lambda x: haversine(x.pickup_latitude,x.pickup_longitude,x.dropoff_latitude,x.dropoff_longitude),axis = 1)

In [26]:
### Haversine median distance
df['haversine_distance'].median()

2.1209923961838655

In [27]:
### Haversine distance maximum
df['haversine_distance'].max()

16409.239135313142

In [29]:
### data entries where haversine distance is 0.0
df[df['haversine_distance']==0.0]

ride_id  fare_amount     pickup_datetime  pickup_longitude  \
5       44470845          4.9 2011-02-12 02:27:09        -73.969019   
7       44195482          2.5 2012-12-11 13:52:00          0.000000   
11       6379048          8.5 2011-05-23 22:15:00          0.000000   
48      22405517         56.8 2013-01-03 22:24:41        -73.993498   
65      21993993          6.0 2014-05-05 19:27:00          0.000000   
...          ...          ...                 ...               ...   
199880  35013970          6.5 2014-02-22 06:45:46          0.000000   
199883  44115598         12.5 2012-09-10 17:39:00          0.000000   
199917  45368488          4.5 2013-06-24 22:17:43        -73.793768   
199932  46517645         24.9 2011-03-22 13:59:00        -73.974618   
199963  50075618         39.0 2012-09-22 07:46:01          0.000000   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
5             40.755910         -73.969019         40.755910                1   
7              0.000000           0.000000          0.000000                1   
11             0.000000           0.000000          0.000000                1   
48            40.764686         -73.993498         40.764686                1   
65             0.000000           0.000000          0.000000                1   
...                 ...                ...               ...              ...   
199880         0.000000           0.000000          0.000000                1   
199883         0.000000           0.000000          0.000000                2   
199917        40.656939         -73.793768         40.656939                1   
199932        40.756295         -73.974618         40.756295                5   
199963         0.000000           0.000000          0.000000                1   

        haversine_distance  
5                      0.0  
7                      0.0  
11                     0.0  
48                     0.0  
65                     0.0  
...                    ...  
199880                 0.0  
199883                 0.0  
199917                 0.0  
199932                 0.0  
199963                 0.0  

[5632 rows x 9 columns]

In [31]:
### mean fare of the minimum distance rides
df[df['haversine_distance']==0.0]['fare_amount'].mean()

11.585317826704546

In [34]:
### distance of the ride with maximum fare
df[df['fare_amount']==df['fare_amount'].max()]['haversine_distance']

170081    0.00079
Name: haversine_distance, dtype: float64

#### Haversine distance calculates the spherical distance between two points. IN the above case where the fare_amount is maximum but the distance is minimum could be because the rider must have started from the pick up location and dropped at the same location after taking his/her journey.

In [35]:
### number of rides in 2014
df[df['pickup_datetime'].dt.year == 2014]

ride_id  fare_amount     pickup_datetime  pickup_longitude  \
4       17610152        16.00 2014-08-28 17:47:00        -73.925023   
6       48725865        24.50 2014-10-12 07:04:00        -73.961447   
20      55085966        10.50 2014-02-18 14:26:00        -73.980022   
26      38755863         5.00 2014-01-21 06:55:00        -73.957802   
34      19277743        39.50 2014-06-04 06:49:00        -73.788080   
...          ...          ...                 ...               ...   
199979   9699676         6.00 2014-06-08 21:38:43        -73.975209   
199981  21553740         9.00 2014-06-02 21:51:50        -73.980445   
199982  13096190        57.33 2014-08-06 11:06:06        -73.969204   
199994   3189201        12.00 2014-01-31 14:42:00        -73.983070   
199996  16382965         7.50 2014-03-14 01:09:00        -73.984722   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
4             40.744085         -73.973082         40.761247                5   
6             40.693965         -73.871195         40.774297                5   
20            40.745990         -74.003432         40.759667                1   
26            40.776372         -73.957422         40.782870                1   
34            40.642187         -73.865042         40.725997                4   
...                 ...                ...               ...              ...   
199979        40.752767         -73.963186         40.757905                2   
199981        40.761823         -73.954004         40.784407                1   
199982        40.754771         -73.790351         40.643802                1   
199994        40.760770         -73.972972         40.754177                1   
199996        40.736837         -74.006672         40.739620                1   

        haversine_distance  
4                 4.475450  
6                11.731015  
20                2.490244  
26                0.723253  
34               11.356141  
...                    ...  
199979            1.162745  
199981            3.356153  
199982           19.483047  
199994            1.122878  
199996            1.875050  

[29968 rows x 9 columns]

In [39]:
### number of rides in first quarter of 2014
df[(df['pickup_datetime'].dt.year == 2014) & ((df['pickup_datetime'].dt.month == 1) | (df['pickup_datetime'].dt.month == 2) | (df['pickup_datetime'].dt.month == 3))]

ride_id  fare_amount     pickup_datetime  pickup_longitude  \
20      55085966         10.5 2014-02-18 14:26:00        -73.980022   
26      38755863          5.0 2014-01-21 06:55:00        -73.957802   
39      38703737         29.0 2014-02-13 17:57:00        -73.992600   
46      37192633         17.0 2014-01-16 14:58:09        -73.993900   
100     29350780          9.0 2014-02-19 18:02:00        -73.958280   
...          ...          ...                 ...               ...   
199931  28516870         11.0 2014-01-08 20:52:14        -74.005809   
199943   9784541         10.0 2014-02-13 18:06:00        -73.970460   
199964   4642171          6.0 2014-02-25 13:37:00        -73.991367   
199994   3189201         12.0 2014-01-31 14:42:00        -73.983070   
199996  16382965          7.5 2014-03-14 01:09:00        -73.984722   

        pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
20            40.745990         -74.003432         40.759667                1   
26            40.776372         -73.957422         40.782870                1   
39            40.753172         -73.908508         40.816192                1   
46            40.751714         -73.958575         40.760390                1   
100           40.768900         -73.973510         40.782907                5   
...                 ...                ...               ...              ...   
199931        40.706212         -73.988523         40.736367                1   
199943        40.759750         -73.984607         40.779352                1   
199964        40.755807         -73.980853         40.761475                2   
199994        40.760770         -73.972972         40.754177                1   
199996        40.736837         -74.006672         40.739620                1   

        haversine_distance  
20                2.490244  
26                0.723253  
39                9.961496  
46                3.127905  
100               2.017541  
...                    ...  
199931            3.655859  
199943            2.483983  
199964            1.086936  
199994            1.122878  
199996            1.875050  

[7687 rows x 9 columns]

In [43]:
### Weekday with maximum number of rides in September 2010
df[(df['pickup_datetime'].dt.year == 2010) & (df['pickup_datetime'].dt.month == 9)]['pickup_datetime'].dt.weekday.max()

6

For the weekday attribute 0 is taken as Monday and 6 as Sunday. Hence, in September 2010, the maximum number of rides occured on Sunday.

#### Model buuilding and finding the best algorithm

In [44]:
df['ride_week_day'] = df['pickup_datetime'].dt.weekday

In [47]:
X,y = df[['passenger_count', 'haversine_distance', 'ride_week_day']],df['fare_amount']

In [49]:
X_train, X_test, y_train, y_test = train_test_split(X,y,
                                                   test_size = 0.3,
                                                   random_state = 42)

#### Linear regression model

In [50]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
reg_model = LinearRegression()
reg_model.fit(X_train,y_train)
y_predicted = reg_model.predict(X_test)
r2_linear = r2_score(y_test,y_predicted)
print(r2_linear)

0.0004341139144697914


#### Decision Tree Classifier won't work for our data as target is continuous. We go with Decision Tree Regressor

#### Decision Tree Regressor

In [53]:
from sklearn.tree import DecisionTreeRegressor
dregressor = DecisionTreeRegressor()
dregressor.fit(X_train, y_train)
y_pred = dregressor.predict(X_test)
r2_dt = r2_score(y_test,y_pred)
print(r2_dt)

0.47398537291521337


#### Random Forest Regressor

In [54]:
from sklearn.ensemble import RandomForestRegressor
rfregressor = RandomForestRegressor(n_estimators = 100)
rfregressor.fit(X_train, y_train)
y_pred = rfregressor.predict(X_test)
r2_rf = r2_score(y_test,y_pred)
print(r2_rf)

0.6300377173534514


#### KNN Classifier won't work for our data as target is continuous. We go with KNN regressor.

#### KNN Regressor

In [55]:
from sklearn import neighbors
# Considering number of neighbours as 10
KNNmodel = neighbors.KNeighborsRegressor(n_neighbors = 10)
KNNmodel.fit(X_train, y_train)
y_pred=KNNmodel.predict(X_test)
r2_knn = r2_score(y_test,y_pred)
print(r2_knn)

0.6557268911518739
